ledger/complete/wal/compactor.go

```diff
@@ -24,48 +27,49 @@ func NewCompactor(checkpointer *Checkpointer, interval time.Duration, checkpoint
 	}
 	return &Compactor{
 		checkpointer:       checkpointer,
-		done:               make(chan struct{}),
 		stopc:              make(chan struct{}),
+		observers:          make(map[observable.Observer]struct{}),
+		lm:                 lifecycle.NewLifecycleManager(),
 		interval:           interval,
 		checkpointDistance: checkpointDistance,
 		checkpointsToKeep:  checkpointsToKeep,
 	}
 }
 
-// Ready periodically fires Run function, every `interval`
-// If called more then once, behaviour is undefined.
-func (c *Compactor) Ready() <-chan struct{} {
-	ch := make(chan struct{})
+func (c *Compactor) Subscribe(observer observable.Observer) {
+	var void struct{}
+	c.observers[observer] = void
+}
 
-	c.wg.Add(1)
-	go c.start()
+func (c *Compactor) Unsubscribe(observer observable.Observer) {
+	delete(c.observers, observer)
+}
 
-	defer close(ch)
-	return ch
+// Ready periodically fires Run function, every `interval`
+func (c *Compactor) Ready() <-chan struct{} {
+	c.lm.OnStart(func() {
+		go c.start()
+	})
+	return c.lm.Started()
 }
 
 func (c *Compactor) Done() <-chan struct{} {
-	c.stopc <- struct{}{}
-
-	ch := make(chan struct{})
-
-	go func() {
-		c.wg.Wait()
-		close(ch)
-	}()
-
-	return ch
+	c.lm.OnStop(func() {
+		for observer := range c.observers {
+			observer.OnComplete()
+		}
+		c.stopc <- struct{}{}
```

Comment:
You may be interested in https://github.com/onflow/flow-go/pull/1077, which will allow you to replace `c.stopc` as well :)